Svc/ComQueue/ComQueue.cpp

```diff
@@ -0,0 +1,308 @@
+    }
+    this->tlmWrite_comQueueDepth(comQueueDepth);
+
+    // Downlink the high-water marks for the Fw::Buffer array types
+    BuffQueueDepth buffQueueDepth;
+    for (FwSizeType i = 0; i < buffQueueDepth.SIZE; i++) {
+        buffQueueDepth[i] = this->m_queues[i + COM_PORT_COUNT].get_high_water_mark();
+        this->m_queues[i + COM_PORT_COUNT].clear_high_water_mark();
+    }
+    this->tlmWrite_buffQueueDepth(buffQueueDepth);
+}
+
+void ComQueue::retryReturn_handler(const NATIVE_INT_TYPE portNum, Fw::Buffer &fwBuffer) {
+    this->m_lock.lock();
+    this->m_bufferRetry = fwBuffer;
+    this->m_lock.unLock();
+}
+
+
+// ----------------------------------------------------------------------
+// Private helper methods
+// ----------------------------------------------------------------------
+
+void ComQueue::enqueue(const FwIndexType queueNum, QueueType queueType, const U8* data, const FwSizeType size) {
+    // Enqueue the given message onto the matching queue. When no space is available then emit the queue overflow event,
+    // set the appropriate throttle, and move on. Will assert if passed a message for a depth 0 queue.
+    const FwSizeType expectedSize = (queueType == QueueType::COM_QUEUE) ? sizeof(Fw::ComBuffer) : sizeof(Fw::Buffer);
+    const FwIndexType portNum = queueNum - ((queueType == QueueType::COM_QUEUE) ? 0 : COM_PORT_COUNT);
+    FW_ASSERT(expectedSize == size, size, expectedSize);
+    FW_ASSERT(portNum >= 0, portNum);
+    Fw::SerializeStatus status = this->m_queues[queueNum].enqueue(data, size);
+    if (status == Fw::FW_SERIALIZE_NO_ROOM_LEFT && !this->m_throttle[queueNum]) {
+        this->log_WARNING_HI_QueueOverflow(queueType, portNum);
+        this->m_throttle[queueNum] = true;
+    }
+    // When the component is already in READY state process the queue to send out the next available message immediately
+    if (this->m_state == READY) {
+        this->processQueue();
+    }
+}
+
+void ComQueue::sendComBuffer(Fw::ComBuffer& comBuffer) {
+    FW_ASSERT(this->m_state == READY);
+    this->m_comRetry = comBuffer;
+    this->comQueueSend_out(0, comBuffer, 0);
+    this->m_state = WAITING;
+}
+
+void ComQueue::sendBuffer(Fw::Buffer& buffer) {
+    FW_ASSERT(this->m_state == READY);
+    this->m_lock.lock();
+    // Buffer ownership transferred to downstream component. Clear local reference.
+    this->m_bufferRetry = Fw::Buffer(nullptr, 0, 0);
+    this->m_lock.unLock();
+    this->buffQueueSend_out(0, buffer);
+    this->m_state = WAITING;
+}
+
+void ComQueue::retryQueue() {
+    // Check that we are in the appropriate state
+    FW_ASSERT(this->m_state == READY);
+
+    // Retry the last message based on type
+    if (this->m_lastIndex < COM_PORT_COUNT) {
+        this->sendComBuffer(this->m_comRetry);
+    } else {
+        // Ensure the retry buffer ownership has been returned before resending it.
+        this->m_lock.lock();
+        FW_ASSERT(this->m_bufferRetry.getData() != nullptr);
```

Comment:
Yes, we could move the clear here.  Clearing it above is a superset as it clears every time a buffer is sent, not just retry.  Since the send of any buffer should result in a new retry buffer it is safe to clear retry in preparation for the callback.

Reply:
I think if we cleared the retry buffer here, then it's more obvious that we should be clearing it, and we won't have a lost buffer.

Reply:
> Since the send of any buffer should result in a new retry buffer it is safe to clear retry in preparation for the callback.

I think this assumes that when we send a buffer B, either we are sending the retry buffer(i.e., B == m_retryBuffer) or there is no retry buffer to send (i.e., m_retryBuffer == NULL). If neither of these conditions holds, then I think we will have a lost buffer.

Reply:
Agreed. I have fixed this.